DRILL-6381: Add support for index based planning and execution

amansinha100 opened the pull request:
- Please see [DRILL-6381](https://issues.apache.org/jira/browse/DRILL-6381) for a description of the feature.  For additional details please see the commit messages.  Wherever appropriate, Javadoc has been provided in the new code. 

- This PR contains 5 commits and has contributions from several people who have been listed as co-authors in individual commits.

- This PR introduces a dependency on MapR 6.0.1 (current apache master is on v5.2.1) which in turn requires CentOS 7.x

amansinha100 commented:
@Ben-Zvi since the code touches HashJoin and a few other run time operators, could you help review the operator changes ? In particular RowKeyJoin, RangePartition, ScanBatch and related changes. 

@arina-ielchiieva do you think you can help review the planner changes and storage plugin changes ? 

Note: All unit and functional tests (including advanced tests) are passing with this branch. 

Thanks in advance..

Ben-Zvi commented on exec/java-exec/src/main/java/org/apache/drill/exec/planner/common/JoinControl.java:
```diff
@@ -0,0 +1,53 @@
+ * bit 0: intersect or not, 0 -- default(no intersect), 1 -- INTERSECT (DISTINCT as default)
+ * bit 1: intersect type, 0 -- default (DISTINCT), 1 -- INTERSECT_ALL
+ */
+public class JoinControl {
```
Looks like the JoinControl currently is not being actually used by the runtime operators; is it just a preparation for future use ?

(A little off the subject)  And it is used to convey information from the planner to the operator. I was recently planning passing other information (e.g., semi-join, anti-semi, broadcast, between lateral and unnest ...) -- any such new bit of information requires so many changes, and extending method signatures; there should be an easier way to do that - like passing the whole plan (but that's another project ....).

amansinha100 replied:
The JoinControl is actually used in IndexIntersectPlanGenerator, NonCoveringIndexPlanGenerator and a few other places.  You may need to expand those files since GitHub does not automatically load the diffs by default. 

For the broader question of passing around the join type and related information in a more componentized way, yes this has been a pending item that planner needs to provide.  Pls feel free to create a JIRA and if you have any initial thoughts you can start proposing it there.

Ben-Zvi replied:
I cherry-picked the 5 commits on top of the (almost) latest master, and searching through the code I could not find who **consumes** the `joinControl` , that is - that value is being passed around in many method calls, and initializes local fields, but no where in the code a **decision** is made based on that value ... 
   Many methods that seem to analyze this value are unused. Like `isIntersect()`, `isIntersectDistinct()`, `isIntersectAll()` in `IndexGroup`,  or `getJoinControl()` in `HashTableConfig`.
    And also unused are the fields `joinControl` in `HashJoinTableTemplate`, or `joinControl` in `HashJoinPrel`.

amansinha100 replied:
Hmm...it seems a small part of the code in HashJoinBatch may not have been merged correctly .. either when I did the recent merge or when previously we merged the index code with the hash join spilling code.  On a separate private branch, I do see `JoinControl` being used for a decision point.  I will talk to you offline about this. 

I am also tagging @chunhui-shi who is a co-author and had added the `JoinControl` specifically to distinguish between INTERSECT_DISTINCT and INTERSECT_ALL.

chunhui-shi replied:
JoinControl has to be used in HashJoin runtime operator, basically it is to allow the operator to skip some operations when it is index join(so we are expecting INTERSECT and nothing else). I don't have the original code at hand, but I think it should be in the first MapR 6.0 release. Since this part is only about performance and correctness won't be impacted, if it was lost during some merge process people might not notice it.

amansinha100 replied:
@Ben-Zvi I have incorporated the joinControl logic as part of the hash join probe phase.  Please see changes in [1].  I haven't done it for build phase since it will be superseded by the semi-join changes.  Can you pls review this change ?

[1] https://github.com/apache/drill/pull/1466/commits/5330fd684a20587e9e860d7894e3d3602a6d0495

HanumathRao commented on contrib/format-maprdb/src/main/java/org/apache/drill/exec/planner/index/MapRDBIndexDiscover.java:
```diff
@@ -0,0 +1,374 @@
+  private DrillIndexDescriptor buildIndexDescriptor(String tableName, IndexDesc desc)
+      throws InvalidIndexDefinitionException {
+    if (desc.isExternal()) {
+      //XX: not support external index
```
Please change this comment to "External index is not supported"

HanumathRao commented on contrib/format-maprdb/src/main/java/org/apache/drill/exec/store/mapr/db/RestrictedMapRDBSubScanSpec.java:
```diff
@@ -0,0 +1,219 @@
+
+      // this is specific to the way the hash join maintains its entries. once we have reached the max
+      // occupied index within a batch, move to the next one and reset the current index to 0
+      // TODO: we should try to abstract this out
```
Please try to simplify this comment.

amansinha100 replied:
This method seems to be a left-over from a previous version of the implementation and it turns out is not needed anymore.  I have removed it.

amansinha100 commented on exec/java-exec/src/main/java/org/apache/drill/exec/planner/index/generators/NonCoveringIndexPlanGenerator.java:
```diff
@@ -0,0 +1,335 @@
+    // index condition. By applying the same index condition again here, we will avoid the possibility to have some
+    //not-wanted records get into downstream operators in such scenarios.
+    //the remainder condition will be applied on top of RowKeyJoin.
+    FilterPrel leftIndexFilterPrel = new FilterPrel(dbScan.getCluster(), dbScan.getTraitSet(),
```
Yes, synchronous indexes (like in RDBMS) obviate the need for re-applying this filter.  I can add an interface method isAsyncIndex() or some such to the IndexCollection and handle this accordingly.. is that ok ?

gparai replied:
Yes, adding it in the interface would be good. We would have to make a similar change in index intersection plan generator as well.

amansinha100 commented on exec/java-exec/src/main/java/org/apache/drill/exec/planner/physical/RowKeyJoinPrel.java:
```diff
@@ -0,0 +1,117 @@
+    }
+    double rowCount = mq.getRowCount(this.getRight());
+    DrillCostFactory costFactory = (DrillCostFactory) planner.getCostFactory();
+    return costFactory.makeCost(rowCount, 0, 0, 0,
```
RowKeyJoin operator by itself incurs negligible CPU and I/O cost since it is not doing a real join.  The actual cost is attributed to the skip-scan (random I/O).  The RK join will hold 1 batch in memory but since it is not a buffering operator, we typically don't assign memory costs to them.   I can add the explanation.

amansinha100 commented on exec/java-exec/src/main/java/org/apache/drill/exec/planner/physical/DrillDistributionTraitDef.java:
```diff
@@ -86,7 +86,10 @@ public RelNode convert(
                                              toDist.getFields());
       case RANGE_DISTRIBUTED:
-        return new OrderedPartitionExchangePrel(rel.getCluster(), planner.emptyTraitSet().plus(Prel.DRILL_PHYSICAL).plus(toDist), rel);
+        // return new OrderedPartitionExchangePrel(rel.getCluster(), planner.emptyTraitSet().plus(Prel.DRILL_PHYSICAL).plus(toDist), rel);
```
The OrderedPartitionExchange was not being used anywhere.  Its original purpose was to do range-based sorting; however this was not completely functional.  It was also a much more complex operator (see its implementation in OrderedPartitionRecordBatch). I think if we go back to fixing it and using it in future, we will probably add a new type of distribution trait.  I will add explanation along these lines.

Ben-Zvi commented on exec/java-exec/src/main/java/org/apache/drill/exec/planner/index/IndexGroup.java:
```diff
@@ -0,0 +1,63 @@
+      return true;
+    } else {
+      return false;
+    }
```
c'mon - why not   `return indexProps.size() > 1 ;` instead ?

amansinha100 replied:
Agreed.  Done.

amansinha100 commented:
@gparai , @HanumathRao I have addressed most of your code review comments.  Working on @Ben-Zvi 's  comment on JoinControl.

vdiravka commented on exec/java-exec/src/main/java/org/apache/drill/exec/planner/index/IndexPlanUtils.java:
```diff
@@ -0,0 +1,872 @@
+   * @return
+   */
+  static public boolean isCoveringIndex(IndexCallContext indexContext, FunctionalIndexInfo functionInfo) {
+    if(functionInfo.hasFunctional()) {
```
space

vdiravka commented on exec/java-exec/src/main/java/org/apache/drill/exec/planner/index/IndexPlanUtils.java:
```diff
@@ -0,0 +1,872 @@
+    //   if references to schema paths in functional indexes disappear beyond capProject
+
+    if (indexContext.getFilter() != null && indexContext.getUpperProject() == null) {
+      if( !isFullQuery(indexContext)) {
```
formatting

vdiravka commented on exec/java-exec/src/main/java/org/apache/drill/exec/planner/index/IndexPlanUtils.java:
```diff
@@ -0,0 +1,872 @@
+    Map<LogicalExpression, Set<SchemaPath>> exprPathMap = functionInfo.getPathsInFunctionExpr();
+    PathInExpr exprSearch = new PathInExpr(exprPathMap);
+
+    for(LogicalExpression expr: exprs) {
```
space

vdiravka commented on exec/java-exec/src/main/java/org/apache/drill/exec/planner/index/IndexPlanUtils.java:
```diff
@@ -0,0 +1,872 @@
+   */
+  private static boolean exprOnlyInEquality(LogicalExpression expr, IndexCallContext context) {
+    //if there is no filter, expr wont be in equality
+    if(context.getFilter() == null) {
```
space

vdiravka commented on exec/java-exec/src/main/java/org/apache/drill/exec/planner/index/IndexPlanUtils.java:
```diff
@@ -0,0 +1,872 @@
+                                  List<LogicalExpression> indexCols,
+                                  IndexConditionInfo.Builder builder, RexNode condition) {
+    boolean prefix = true;
+    int i=0;
```
spaces

vdiravka commented on exec/java-exec/src/main/java/org/apache/drill/exec/planner/index/IndexPlanUtils.java:
```diff
@@ -0,0 +1,872 @@
+      LogicalExpression p = indexCols.get(i++);
+      List<LogicalExpression> prefixCol = ImmutableList.of(p);
+      IndexConditionInfo info = builder.indexConditionRelatedToFields(prefixCol, initCondition);
+      if(info != null && info.hasIndexCol) {
```
space

vdiravka commented on exec/java-exec/src/main/java/org/apache/drill/exec/planner/index/IndexSelector.java:
```diff
@@ -0,0 +1,766 @@
+              indexContext.getScan(), functionInfo, indexContext);
+    }
+
+    if ( (inputCollation != null) &&
```
formatting

vdiravka commented on protocol/src/main/java/org/apache/drill/exec/proto/beans/CoreOperatorType.java:
```diff
@@ -78,7 +78,8 @@
     PCAPNG_SUB_SCAN(55),
-    RUNTIME_FILTER(56);
+    RUNTIME_FILTER(56),
+    ROWKEY_JOIN(57);
```
Please regenerate protobuf files for C++ native client as well

amansinha100 replied:
I have regenerated the protobuf files for the native C++ client.  (As an aside, the steps for doing this seems complicated unlike the Java version. In either case, we really should not need to rebuild the client side if a new operator has been added on the server side).

vdiravka commented on pom.xml:
```diff
@@ -53,8 +53,8 @@
     <zookeeper.version>3.4.12</zookeeper.version>
-    <mapr.release.version>5.2.1-mapr</mapr.release.version>
-    <ojai.version>1.1</ojai.version>
+    <mapr.release.version>6.0.1-mapr</mapr.release.version>
```
Use 6.1.0-mapr version
It will be introduced after merging this PR:
https://github.com/apache/drill/pull/1489/files#diff-600376dffeb79835ede4a0b285078036R56

amansinha100 replied:
For this version update, I will let the the PR you referenced [1] handle it.  There are functional test failures with that PR that need to be addressed. 

[1] https://github.com/apache/drill/pull/1489

vdiravka commented on exec/java-exec/src/main/java/org/apache/drill/exec/planner/physical/PlannerSettings.java:
```diff
@@ -114,6 +114,28 @@
+  public static final String USE_SIMPLE_OPTIMIZER_KEY = "planner.use_simple_optimizer";
+  public static final BooleanValidator USE_SIMPLE_OPTIMIZER = new BooleanValidator(USE_SIMPLE_OPTIMIZER_KEY, null);
+  public static final BooleanValidator INDEX_PLANNING = new BooleanValidator("planner.enable_index_planning", null);
+  public static final BooleanValidator ENABLE_STATS = new BooleanValidator("planner.enable_statistics", null);
```
Is it used for obtaining MapR-DB and Parquet table statistics?

amansinha100 replied:
I will let @gparai elaborate on this but my understanding is that by keeping a general name such as `planner.enable_statistics` option we could potentially use it for any type of data source (including Parquet) once the broader stats capability is added to Drill.

gparai replied:
Yes, that was the intent - allow statistics use for any source with one option.

vdiravka commented on exec/java-exec/src/main/java/org/apache/drill/exec/planner/logical/DrillOptiq.java:
```diff
@@ -22,7 +22,7 @@
 import java.util.List;
 
-import org.apache.drill.shaded.guava.com.google.common.base.Preconditions;
+import org.apache.calcite.rel.type.RelDataType;
```
Optiq is an old name of Calcite. Does it make sense to rename this class?

amansinha100 replied:
It probably does..can you create a JIRA for the renaming since it is independent of this PR.

vdiravka commented on exec/java-exec/src/main/java/org/apache/drill/exec/planner/index/InvalidIndexDefinitionException.java:
```diff
@@ -0,0 +1,27 @@
+
+import org.apache.drill.common.exceptions.DrillRuntimeException;
+
+public class InvalidIndexDefinitionException extends DrillRuntimeException {
```
Consider adding some description to the exception

amansinha100 replied:
Done

vdiravka commented on exec/java-exec/src/main/java/org/apache/drill/exec/planner/index/SimpleRexRemap.java:
```diff
@@ -0,0 +1,300 @@
+        }
+        else if (operands.get(1) instanceof RexLiteral) {
+          oplit = (RexLiteral) operands.get(1);
+          if(oplit.getTypeName() == SqlTypeName.CHAR) {
```
space

vdiravka commented on exec/java-exec/src/main/java/org/apache/drill/exec/planner/index/generators/IndexIntersectPlanGenerator.java:
```diff
@@ -0,0 +1,350 @@
+    Pair<RelNode, DbGroupScan> leftRelAndScan = buildRestrictedDBScan(remnant, isAnyIndexAsync);
+
+    RelNode finalRel = buildRowKeyJoin(leftRelAndScan.left, rangeDistRight, true, JoinControl.DEFAULT);
+    if ( upperProject != null) {
```
formatting

vdiravka commented on exec/java-exec/src/main/java/org/apache/drill/exec/planner/index/rules/AbstractMatchFunction.java:
```diff
@@ -0,0 +1,58 @@
+    }
+    return false;
+  }
+}
```
new line

vdiravka commented on exec/java-exec/src/main/java/org/apache/drill/exec/planner/index/rules/MatchFunction.java:
```diff
@@ -0,0 +1,25 @@
+public interface MatchFunction<T> {
+  boolean match(RelOptRuleCall call);
+  T onMatch(RelOptRuleCall call);
+}
```
new line

vdiravka commented on exec/java-exec/src/main/java/org/apache/drill/exec/planner/logical/DrillMergeProjectRule.java:
```diff
@@ -166,4 +169,25 @@ public void onMatch(RelOptRuleCall call) {
+        RelOptUtil.pushPastProject(topProject.getProjects(), bottomProject);
+
+    // replace the two projects with a combined projection
+    if(topProject instanceof DrillProjectRel) {
```
space

vdiravka commented on exec/java-exec/src/main/java/org/apache/drill/exec/planner/logical/DrillMergeProjectRule.java:
```diff
@@ -166,4 +169,25 @@ public void onMatch(RelOptRuleCall call) {
     return list;
   }
 
+  public static Project replace(Project topProject, Project bottomProject) {
```
Can `ProjectAllowDup` and `Project` rels be merged?
For instance:
`00-01      ProjectAllowDup(MONTHID=[$0], NOFOCONTRACTS=[$1]) : rowType = RecordType(INTEGER MONTHID, INTEGER NOFOCONTRACTS)`
`00-02        Project(MonthId=[CAST($1):INTEGER], No of Contracts=[CAST($2):INTEGER]) : rowType = RecordType(INTEGER MonthId, INTEGER No of Contracts)`

amansinha100 replied:
They can be merged into the top level Project that allows duplicates...however, the purpose of the `replace()` method here is to simply allow the caller to replace a project with another with the assumption that callers know exactly what they are doing.  This is not applying the full fledged `DrillMergeProjectRule`.

vdiravka replied:
It makes sense. A small javadoc can be useful here.

amansinha100 replied:
Added javadoc.

amansinha100 commented on contrib/format-maprdb/src/main/java/org/apache/drill/exec/store/mapr/db/json/MaprDBJsonRecordReader.java:
```diff
@@ -517,6 +388,15 @@ private static FieldPath getFieldPathForProjection(SchemaPath column) {
   }
 
+  public static boolean includesIdField(Collection<FieldPath> projected) {
+    return Iterables.tryFind(projected, new Predicate<FieldPath>() {
```
@vdiravka could you elaborate on this a bit ?  I haven't done much with lambda expressions in Java 8, but if you show how to rewrite this statement and assuming it has an advantage over the existing implementation, I would be happy to.  Thanks.

amansinha100 replied:
Since this is a syntactic modification, I am inclined to leave it as-is .. partly because I might cause an unintentional side effect/regression and partly because such functional utilities appear elsewhere in the Drill and Calcite code.  Ideally, if we want to move to Lambdas, we should re-visit all such implementations.

vdiravka replied:
Agreed. 
Vova was doing that, but has not finished in the result.

amansinha100 commented:
@vdiravka I think I have addressed all the formatting issues in commit 5e0fa72 and added some missing javadocs (btw, thanks for pointing out the formatting issues..I don't think I introduced them but    should have caught them before creating the PR).  I will go ahead and 'hide' your comments related to these and leave your other comments open.

amansinha100 commented on exec/java-exec/src/main/java/org/apache/drill/exec/planner/index/DrillIndexDescriptor.java:
```diff
@@ -0,0 +1,110 @@
+    return null;
+  }
+
+  public void attach(String storageName, DrillTable inTable) {
```
Removed attach() since it was not being used.  Added Javadoc or `@Override` for few others.

amansinha100 commented on contrib/format-maprdb/src/main/java/org/apache/drill/exec/store/mapr/db/MapRDBPushProjectIntoScan.java:
```diff
@@ -0,0 +1,145 @@
+
+import java.util.List;
+
+public abstract class MapRDBPushProjectIntoScan extends StoragePluginOptimizerRule {
```
The main difference is that DrillPushProjectIntoScanRule is applied during logical planning phase whereas the MapRDBPushProjectIntoScan is applied during physical planning because we want to ensure that even after new physical plans are created (such as index based plans), the projection pushdown is done.

vdiravka replied:
It makes sense, thank you for clarifications.

Ben-Zvi commented on exec/java-exec/src/main/java/org/apache/drill/exec/planner/index/generators/IndexIntersectPlanGenerator.java:
```diff
@@ -0,0 +1,350 @@
+        RelOptUtil.createEquiJoinCondition(left, leftJoinKeys,
+            right, rightJoinKeys, builder);
+
+    if (isRowKeyJoin == true) {
```
How about `if (isRowKeyJoin) {` instead ?

amansinha100 replied:
done

vdiravka commented on contrib/format-maprdb/src/main/java/org/apache/drill/exec/planner/index/MapRDBFunctionalIndexInfo.java:
```diff
@@ -0,0 +1,168 @@
+
+  public MapRDBFunctionalIndexInfo(IndexDescriptor indexDesc) {
+    this.indexDesc = indexDesc;
+    columnToConvert = Maps.newHashMap();
```
You can use it, but from the last Guava version used by Drill (23.0 version) there is a note for this factory method:
> <p><b>Note for Java 7 and later:</b> this method is now unnecessary and
> should be treated as deprecated. Instead, use the {@code HashMap}
> constructor directly, taking advantage of the new
> <a href="http://goo.gl/iz2Wi">"diamond" syntax</a>.
https://google.github.io/guava/releases/snapshot/api/docs/com/google/common/collect/Maps.html#newHashMap--

There was no this note in the previous Guava version used by Drill (19.0 version), therefore some usages of this method are left.

amansinha100 replied:
The checkstyle-validation had identified a bunch of non-shaded Guava classes in the core classes when I created the PR and I had fixed those by using the shaded version. But oddly, it did not complain for the MapR-DB plugin classes. Does this enforcement not occur for the plugins ? I am not sure..but in any case I looked through the new code added in MapR-DB plugin and have modified the imports to use shaded Guava classes.

amansinha100 replied:
Ah, that explains it.  Presumably we will fix it for 1.15.  Thanks.

amansinha100 commented on contrib/format-maprdb/src/main/java/org/apache/drill/exec/store/mapr/db/MapRDBPushLimitIntoScan.java:
```diff
@@ -0,0 +1,203 @@
+import org.apache.drill.exec.store.mapr.db.json.JsonTableGroupScan;
+import org.apache.drill.exec.store.mapr.db.json.RestrictedJsonTableGroupScan;
+
+public abstract class MapRDBPushLimitIntoScan extends StoragePluginOptimizerRule {
```
The `DrillPushLimitToScanRule` works for the logical plan.  We introduced the MapR-DB specific rule to apply similar pushdown for the physical plan because after index planning is done additional pushdowns may be possible.

amansinha100 commented on contrib/format-maprdb/src/main/java/org/apache/drill/exec/store/mapr/db/MapRDBPushProjectIntoScan.java:
```diff
@@ -0,0 +1,145 @@
+        return;
+      }
+      doPushProjectIntoGroupScan(call, project, scan, (MapRDBGroupScan) scan.getGroupScan());
+      if (scan.getGroupScan() instanceof BinaryTableGroupScan) {
```
I have made the changes here and rest of this class to only check for JsonTableGroupScan.  Binary table's projection pushdown was already handled by the logical planning rule `DrillPushProjectIntoScan`.   In future, if we want to extend this rule to binary tables, we would need to ensure it is tested.

amansinha100 commented on contrib/format-maprdb/src/main/java/org/apache/drill/exec/store/mapr/db/MapRDBPushFilterIntoScan.java:
```diff
@@ -187,7 +186,7 @@ protected void doPushFilterIntoBinaryGroupScan(final RelOptRuleCall call,
     newGroupsScan.setFilterPushedDown(true);
 
-    final ScanPrel newScanPrel = ScanPrel.create(scan, filter.getTraitSet(), newGroupsScan, scan.getRowType());
+    final ScanPrel newScanPrel = ScanPrel.create(scan, filter.getTraitSet(), newGroupsScan, scan.getRowType(), scan.getTable());
```
Changed to use the constructor.  @vdiravka why not remove the `create` method altogether ?

amansinha100 commented on contrib/format-maprdb/src/main/java/org/apache/drill/exec/store/mapr/db/MapRDBFormatPlugin.java:
```diff
@@ -57,34 +64,74 @@ public MapRDBFormatPlugin(String name, DrillbitContext context, Configuration fs
     connection = ConnectionFactory.createConnection(hbaseConf);
+    jsonTableCache = new MapRDBTableCache(context.getConfig());
+    int scanRangeSizeMBConfig = context.getConfig().getInt(PluginConstants.JSON_TABLE_SCAN_SIZE_MB);
+    if (scanRangeSizeMBConfig < 32 || scanRangeSizeMBConfig > 8192) {
```
Done

amansinha100 commented on contrib/format-maprdb/src/main/java/org/apache/drill/exec/store/mapr/db/MapRDBPushFilterIntoScan.java:
```diff
@@ -137,11 +137,10 @@ protected void doPushFilterIntoJsonGroupScan(RelOptRuleCall call,
 
-    // clone the groupScan with the newScanSpec.
-    final JsonTableGroupScan newGroupsScan = groupScan.clone(newScanSpec);
+    final JsonTableGroupScan newGroupsScan = (JsonTableGroupScan) groupScan.clone(newScanSpec);
```
Currently, the `clone()` method returns an instance of `GroupScan`, so this cast is needed. Did you mean change the original `clone` implementation itself ?  That will require some additional changes since both binary table and json table implement it.

amansinha100 commented on contrib/format-maprdb/src/test/java/com/mapr/drill/maprdb/tests/index/IndexPlanTest.java:
```diff
@@ -0,0 +1,1715 @@
+            " where t1.driverlicense > 100000001 group by t1.driverlicense limit 2";
+    try {
+      test(defaultHavingIndexPlan);
+      test(disableHashAgg);
```
@vdiravka since this is a fairly large test file (close to 75 tests) this suggestion and couple of other suggestions for the test changes seem nice-to-have but perhaps not must-have ?  In the interest of getting the main code changes merged, would it be okay to defer this ?

vdiravka replied:
Surely, there are a lot of places in Drill unit tests, which can be refactored. At least `BaseTestQuery` is marked as `@Deprecated`, but is widely used.
Possibly we can gather all this small issues and make the global refactoring in the process of some task, for example in the process of updating JUnit to 5 version.

amansinha100 commented on exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/partitionsender/Partitioner.java:
```diff
@@ -29,6 +29,8 @@
 import org.apache.drill.exec.record.RecordBatch;
 
 public interface Partitioner {
+  int DEFAULT_RECORD_BATCH_SIZE = (1 << 10) - 1;
```
The DEFAULT_RECORD_BATCH_SIZE is already in the current master branch here [1].  I simply moved it into the `Partitioner` interface. 

[1] https://github.com/apache/drill/blob/master/exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/partitionsender/PartitionerTemplate.java#L58

vdiravka replied:
I see, thanks for clarification.
Is that comment regarding `recordCount` isn't necessary anymore?

amansinha100 replied:
I think that was an omission.  I have re-inserted the comment but with some modifications because the batch-sizing project now allows operators to set the output batch size in terms of Mbytes rather than `recordCount`.  It is not yet applied across the board, so this `DEFAULT_RECORD_BATCH_SIZE` is still relevant for the exchange operators.

chunhui-shi commented:
+1. I think you already addressed all the comments and just need to solve the rebase conflicts.

Ben-Zvi commented on exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/join/HashJoinBatch.java:
```diff
@@ -568,25 +603,13 @@ public IterOutcome innerNext() {
-            leftUpstream = next(HashJoinHelper.LEFT_INPUT, probeBatch);
-          }
-        }
+        drainLeft();
```
Note the expected merge conflict here (from DRILL-6755 = #1480):  The new code defined similar methods (e.g. `killAndDrainLeftUpstream()`). So please follow the #1480 new code.

amansinha100 replied:
I have resolved this merge conflict after rebasing on my local branch.  I haven't pushed it remote yet (need to see if I can avoid force-push).

Ben-Zvi commented on exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/join/HashJoinBatch.java:
```diff
@@ -1212,6 +1238,81 @@ private void updateStats() {
     this.stats.setLongStat(Metric.SPILLED_PARTITIONS, numSpilled);
   }
 
+  private void drainLeft() {
```
Follow on (after #1480) - these two methods can be eliminated ( `drainLeft()` and `drainRight()` )

amansinha100 commented on exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/mergereceiver/MergingRecordBatch.java:
```diff
@@ -535,7 +535,10 @@ public FragmentContext getContext() {
+    if (outgoingContainer.hasSchema()) {
+      return outgoingContainer.getSchema();
+    }
+    return null;
```
The null value could occur for example when even though the `VectorContainer` has vectors but the operator has not yet called `buildSchema()` on the container.  There's a `Preconditions.checkNotNull` in `VectorContainer.getSchema()`.

amansinha100 commented on exec/java-exec/src/main/java/org/apache/drill/exec/planner/cost/DrillRelMdSelectivity.java:
```diff
@@ -0,0 +1,90 @@
+  public static final RelMetadataProvider SOURCE = ReflectiveRelMetadataProvider.reflectiveSource(BuiltInMethod.SELECTIVITY.method, INSTANCE);
+
+
+  public Double getSelectivity(RelNode rel, RexNode predicate) {
```
There were customizations needed for RelSubset, DrillScanRel and ScanPrel selectivities.  For all other nodes, we are calling the `super.getSelectivity()`.

amansinha100 commented on contrib/format-maprdb/src/main/java/org/apache/drill/exec/store/mapr/db/MapRDBPushProjectIntoScan.java:
```diff
@@ -0,0 +1,145 @@
+
+      if (ProjectRemoveRule.isTrivial(newProj) &&
+          // the old project did not involve any column renaming
+          sameRowTypeProjectionsFields(project.getRowType(), newScan.getRowType())) {
```
This may have to do with the fact that `MapRDBPushProjectIntoScan` is applied during physical planning unlike the `DrillPushProjectIntoScanRule` so there may be some differences in how the pushdown occurs.  We would have added the extra check `sameRowTypeProjectionsFields` to prevent an incorrect pushdown, although I don't recall exact details now.

amansinha100 commented on exec/java-exec/src/main/java/org/apache/drill/exec/planner/physical/BroadcastExchangePrel.java:
```diff
@@ -57,7 +57,9 @@ public RelOptCost computeSelfCost(RelOptPlanner planner, RelMetadataQuery mq) {
-    final double networkCost = broadcastFactor * DrillCostBase.BYTE_NETWORK_COST * inputRows * rowWidth * numEndPoints;
+
+    //we assume localhost network cost is 1/10 of regular network cost
+    final double networkCost = broadcastFactor * DrillCostBase.BYTE_NETWORK_COST * inputRows * rowWidth * (numEndPoints - 0.9);
```
@gparai,  forgot to respond to this.  The cost formula is: 
    (cost of broadcasting num_bytes to N - 1 nodes)  +  (cost of local broadcast to all minor fragments on my own node)
   = (C * num_bytes * (N - 1) )  +  (C * num_bytes * 0.1)   where the 0.1 factor comes from the assumption that local broadcast is 10% network cost of the remote broadcast. 
  = C * num_bytes * (N - 0.9)

While the formula seems reasonable, it is biasing the cost in favor of Broadcast compared to HashPartition.  We should re-visit this and ideally a similar change should be done for HashPartition also.  I don't recall the exact use case which motivated the change.. it may have been the Index Intersection.  I can create a JIRA to re-visit this.

amansinha100 replied:
I have added comments in the code regarding this.

amansinha100 commented:
I will close this PR and open a separate one that is rebased on apache master.  The reason is after rebase and squashing some commits, I am having to force-push but that may clobber some of the review comments which I want to avoid.

amansinha100 commented:
Opened the new PR:  https://github.com/apache/drill/pull/1512   to do the final merge.
